nixos/sssd: Add secrets handling (patch originally from @yayayayaka)

Ma27 opened the pull request:
###### Description of changes

Basically #139466 from @yayayayaka, but they closed it because they don't needed it anymore. For me, the change became useful quite recently and thus I'd like to see it in the nixpkgs master. Also, while there are perhaps even nicer solutions, I don't think that this justifies delaying a fix which helps to move secrets out of the Nix store!

Currently, it is not possible to supply sensitive credentials like
`ldap_default_authtok` without writing them to the nix store. This
This commit introduces a new option `environmentFile` where those
credentials can be supplied via environment substitution.


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

yayayayaka committed:
nixos/sssd: Add secrets handling

Currently, it is not possible to supply sensitive credentials like
`ldap_default_authtok` without writing them to the nix store. This
This commit introduces a new option `environmentFile` where those
credentials can be supplied via environment substitution.

RaitoBezarius commented on nixos/modules/services/misc/sssd.nix:
```diff
@@ -3,6 +3,10 @@ with lib;
+
+  dataDir = "/var/lib/sssd";
+  settingsFile = "${dataDir}/sssd.conf";
+  settingsFileUnsubstituted = pkgs.writeText "${dataDir}/sssd-unsubsituted.conf" cfg.config;
```
typo on the actual filename: unsubstituted

RaitoBezarius commented on nixos/modules/services/misc/sssd.nix:
```diff
@@ -60,22 +88,28 @@ in {
           Type = "forking";
           PIDFile = "/run/sssd.pid";
+          StateDirectory = baseNameOf dataDir;
+          EnvironmentFile = lib.mkIf (cfg.environmentFile != null) cfg.environmentFile;
```
Shouldn't this use `LoadCredential` rather?

Ma27 replied:
LoadCredential doesn't work with ExecStartPre. Also, sssd doesn't just have LDAP support so we need a generic solution with envsubst to be able to substitute any credentials.

RaitoBezarius replied:
I see @ ExecStartPre.

RaitoBezarius replied:
Maybe a comment would be great.

RaitoBezarius commented on nixos/modules/services/misc/sssd.nix:
```diff
@@ -60,22 +88,28 @@ in {
         serviceConfig = {
           Type = "forking";
           PIDFile = "/run/sssd.pid";
+          StateDirectory = baseNameOf dataDir;
```
How does that work when `dataDir` is any path which does not start with `/var/lib`, e.g. `dataDir = /data/sssd`, it would give a `StateDirectory = "sssd";` and creates a folder in `/var/lib/sssd`, right?

Ma27 replied:
`dataDir` is hardcoded and not an option.

RaitoBezarius replied:
Ha, I stand corrected then.

Ma27 committed:
nixos/sssd: fix typo

Ma27 committed:
nixos/sssd-ldap: fix eval

Introduced by fd7d901133f9fbfc893cdb33f7d630846bb21f9c. The openldap
module now expects the database directory to be below
`/var/lib/openldap`, oterhwise it'll fail evaluation like this:

    Failed assertions:
    - Database dc=example,dc=org has `olcDbDirectory` (/var/db/openldap) that is not a subdirectory of
    `/var/lib/openldap/`.

Ma27 committed:
nixos/sssd-ldap: verify that passing secrets via env works

Ma27 committed:
nixos/sssd: explain why we use EnvironmentFile=

RaitoBezarius reviewed (approved):
LGTM.
